{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fc85cfea",
   "metadata": {},
   "source": [
    "# Reading in data with help of lymphoseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "99e658a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import lymphoseq as ls\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import polars as pl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "ddc3fbca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>LOC</th>\n",
       "      <th>cells</th>\n",
       "      <th>newname</th>\n",
       "      <th>shortname</th>\n",
       "      <th>genotype_short</th>\n",
       "      <th>sample_short</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FIN_15625_AgXP_TCRB.tsv.gz</td>\n",
       "      <td>/Users/ania/Documents/DQ2DQ8/data_link/D202102...</td>\n",
       "      <td>E</td>\n",
       "      <td>FIN_15625</td>\n",
       "      <td>F15625</td>\n",
       "      <td>heteroDQ2DQ8</td>\n",
       "      <td>F15625E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FIN_15625_NAIVE_TCRB.tsv.gz</td>\n",
       "      <td>/Users/ania/Documents/DQ2DQ8/data_link/D202102...</td>\n",
       "      <td>N</td>\n",
       "      <td>FIN_15625</td>\n",
       "      <td>F15625</td>\n",
       "      <td>heteroDQ2DQ8</td>\n",
       "      <td>F15625N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FIN_16018_AgXP_TCRB.tsv.gz</td>\n",
       "      <td>/Users/ania/Documents/DQ2DQ8/data_link/D202102...</td>\n",
       "      <td>E</td>\n",
       "      <td>FIN_16018</td>\n",
       "      <td>F16018</td>\n",
       "      <td>homoDQ8</td>\n",
       "      <td>F16018E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FIN_16018_NAIVE_TCRB.tsv.gz</td>\n",
       "      <td>/Users/ania/Documents/DQ2DQ8/data_link/D202102...</td>\n",
       "      <td>N</td>\n",
       "      <td>FIN_16018</td>\n",
       "      <td>F16018</td>\n",
       "      <td>homoDQ8</td>\n",
       "      <td>F16018N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FIN_18072_AgXP_TCRB.tsv.gz</td>\n",
       "      <td>/Users/ania/Documents/DQ2DQ8/data_link/D202102...</td>\n",
       "      <td>E</td>\n",
       "      <td>FIN_18072</td>\n",
       "      <td>F18072</td>\n",
       "      <td>heteroDQ2DQ8</td>\n",
       "      <td>F18072E</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        SAMPLE  \\\n",
       "0   FIN_15625_AgXP_TCRB.tsv.gz   \n",
       "1  FIN_15625_NAIVE_TCRB.tsv.gz   \n",
       "2   FIN_16018_AgXP_TCRB.tsv.gz   \n",
       "3  FIN_16018_NAIVE_TCRB.tsv.gz   \n",
       "4   FIN_18072_AgXP_TCRB.tsv.gz   \n",
       "\n",
       "                                                 LOC cells    newname  \\\n",
       "0  /Users/ania/Documents/DQ2DQ8/data_link/D202102...     E  FIN_15625   \n",
       "1  /Users/ania/Documents/DQ2DQ8/data_link/D202102...     N  FIN_15625   \n",
       "2  /Users/ania/Documents/DQ2DQ8/data_link/D202102...     E  FIN_16018   \n",
       "3  /Users/ania/Documents/DQ2DQ8/data_link/D202102...     N  FIN_16018   \n",
       "4  /Users/ania/Documents/DQ2DQ8/data_link/D202102...     E  FIN_18072   \n",
       "\n",
       "  shortname genotype_short sample_short  \n",
       "0    F15625   heteroDQ2DQ8      F15625E  \n",
       "1    F15625   heteroDQ2DQ8      F15625N  \n",
       "2    F16018        homoDQ8      F16018E  \n",
       "3    F16018        homoDQ8      F16018N  \n",
       "4    F18072   heteroDQ2DQ8      F18072E  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs_to_remove = \"data/D20210208D_1-overrepresented-sequences.txt\"\n",
    "output_loc = \"results_man/cleaned\"\n",
    "samplesheet = pd.read_csv(\"data/collated_info.csv\")\n",
    "samplesheet.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "8be5fe58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                          nucleotide          aminoAcid  \\\n",
      "0  CACCTACACACCCTGCAGCCAGAAGACTCGGCCCTGTATCTCTGCG...      CASSQDKNQPQHF   \n",
      "1  ATCCAGCGCACACAGCAGGAGGACTCGGCCGTGTATCTCTGTGCCA...     CASSLMSGPGELFF   \n",
      "2  TCAGAACCCAGGGACTCAGCTGTGTACTTCTGTGCCAGCAGACGCC...  CASRRPTGQGIRSGYTF   \n",
      "3  ATCAATTCCCTGGAGCTTGGTGACTCTGCTGTGTATTTCTGTGCCA...     CASSQGTGGSPLHF   \n",
      "4  GATCCAGCGCACAGAGCAGGGGGACTCAGCTGTGTATCTCTGTGCC...                NaN   \n",
      "\n",
      "   count (templates/reads)  frequencyCount (%)  cdr3Length  \\\n",
      "0                       14            0.000759          39   \n",
      "1                        5            0.000271          42   \n",
      "2                        9            0.000488          51   \n",
      "3                        8            0.000434          42   \n",
      "4                        4            0.000217          41   \n",
      "\n",
      "          vMaxResolved vFamilyName         vGeneName  vGeneAllele vFamilyTies  \\\n",
      "0        TCRBV04-03*01     TCRBV04        TCRBV04-03          1.0         NaN   \n",
      "1           TCRBV07-02     TCRBV07        TCRBV07-02          NaN         NaN   \n",
      "2              TCRBV12     TCRBV12               NaN          NaN         NaN   \n",
      "3  TCRBV03-01/03-02*01     TCRBV03  TCRBV03-01/03-02          1.0         NaN   \n",
      "4        TCRBV07-04*01     TCRBV07        TCRBV07-04          1.0         NaN   \n",
      "\n",
      "                 vGeneNameTies vGeneAlleleTies   dMaxResolved dFamilyName  \\\n",
      "0                          NaN             NaN            NaN         NaN   \n",
      "1                          NaN           01,02     TCRBD02-01     TCRBD02   \n",
      "2  TCRBV12-03/12-04,TCRBV12-04             NaN  TCRBD01-01*01     TCRBD01   \n",
      "3                          NaN             NaN  TCRBD01-01*01     TCRBD01   \n",
      "4                          NaN             NaN  TCRBD01-01*01     TCRBD01   \n",
      "\n",
      "    dGeneName  dGeneAllele      dFamilyTies          dGeneNameTies  \\\n",
      "0         NaN          NaN  TCRBD01,TCRBD02  TCRBD01-01,TCRBD02-01   \n",
      "1  TCRBD02-01          NaN              NaN                    NaN   \n",
      "2  TCRBD01-01          1.0              NaN                    NaN   \n",
      "3  TCRBD01-01          1.0              NaN                    NaN   \n",
      "4  TCRBD01-01          1.0              NaN                    NaN   \n",
      "\n",
      "  dGeneAlleleTies   jMaxResolved jFamilyName   jGeneName  jGeneAllele  \\\n",
      "0             NaN  TCRBJ01-05*01     TCRBJ01  TCRBJ01-05          1.0   \n",
      "1           01,02  TCRBJ02-02*01     TCRBJ02  TCRBJ02-02          1.0   \n",
      "2             NaN  TCRBJ01-02*01     TCRBJ01  TCRBJ01-02          1.0   \n",
      "3             NaN  TCRBJ01-06*02     TCRBJ01  TCRBJ01-06          2.0   \n",
      "4             NaN  TCRBJ01-02*01     TCRBJ01  TCRBJ01-02          1.0   \n",
      "\n",
      "   jFamilyTies jGeneNameTies jGeneAlleleTies  vDeletion  n1Insertion  \\\n",
      "0          NaN           NaN             NaN          0            2   \n",
      "1          NaN           NaN             NaN          2            4   \n",
      "2          NaN           NaN             NaN          6            9   \n",
      "3          NaN           NaN             NaN          1            2   \n",
      "4          NaN           NaN             NaN          4            3   \n",
      "\n",
      "   d5Deletion  d3Deletion  n2Insertion  jDeletion  vIndex  n1Index  dIndex  \\\n",
      "0           5           5            0          4      42       59      61   \n",
      "1           8           4            2          6      39       54      58   \n",
      "2           0           6           13          8      30       41      50   \n",
      "3           3           1            0          9      39       55      57   \n",
      "4           2           5            3          3      40       53      56   \n",
      "\n",
      "   n2Index  jIndex  estimatedNumberGenomes sequenceStatus cloneResolved  \\\n",
      "0       -1      63                      14             In           VDJ   \n",
      "1       62      64                       5             In           VDJ   \n",
      "2       56      69                       9             In           VDJ   \n",
      "3       -1      65                       8             In           VDJ   \n",
      "4       61      64                       4            Out           VDJ   \n",
      "\n",
      "   vOrphon  dOrphon  jOrphon  vFunction  dFunction  jFunction  \\\n",
      "0      NaN      NaN      NaN        NaN        NaN        NaN   \n",
      "1      NaN      NaN      NaN        NaN        NaN        NaN   \n",
      "2      NaN      NaN      NaN        NaN        NaN        NaN   \n",
      "3      NaN      NaN      NaN        NaN        NaN        NaN   \n",
      "4      NaN      NaN      NaN        NaN        NaN        NaN   \n",
      "\n",
      "   fractionNucleated  vAlignLength  vAlignSubstitutionCount  \\\n",
      "0                NaN           NaN                      NaN   \n",
      "1                NaN           NaN                      NaN   \n",
      "2                NaN           NaN                      NaN   \n",
      "3                NaN           NaN                      NaN   \n",
      "4                NaN           NaN                      NaN   \n",
      "\n",
      "   vAlignSubstitutionIndexes  vAlignSubstitutionGeneThreePrimeIndexes  \\\n",
      "0                        NaN                                      NaN   \n",
      "1                        NaN                                      NaN   \n",
      "2                        NaN                                      NaN   \n",
      "3                        NaN                                      NaN   \n",
      "4                        NaN                                      NaN   \n",
      "\n",
      "   vSeqWithMutations sample_short  \n",
      "0                NaN      F15625E  \n",
      "1                NaN      F15625E  \n",
      "2                NaN      F15625E  \n",
      "3                NaN      F15625E  \n",
      "4                NaN      F15625E  \n",
      "['nucleotide', 'aminoAcid', 'count (templates/reads)', 'frequencyCount (%)', 'cdr3Length', 'vMaxResolved', 'vFamilyName', 'vGeneName', 'vGeneAllele', 'vFamilyTies', 'vGeneNameTies', 'vGeneAlleleTies', 'dMaxResolved', 'dFamilyName', 'dGeneName', 'dGeneAllele', 'dFamilyTies', 'dGeneNameTies', 'dGeneAlleleTies', 'jMaxResolved', 'jFamilyName', 'jGeneName', 'jGeneAllele', 'jFamilyTies', 'jGeneNameTies', 'jGeneAlleleTies', 'vDeletion', 'n1Insertion', 'd5Deletion', 'd3Deletion', 'n2Insertion', 'jDeletion', 'vIndex', 'n1Index', 'dIndex', 'n2Index', 'jIndex', 'estimatedNumberGenomes', 'sequenceStatus', 'cloneResolved', 'vOrphon', 'dOrphon', 'jOrphon', 'vFunction', 'dFunction', 'jFunction', 'fractionNucleated', 'vAlignLength', 'vAlignSubstitutionCount', 'vAlignSubstitutionIndexes', 'vAlignSubstitutionGeneThreePrimeIndexes', 'vSeqWithMutations', 'sample_short']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/bj/8722bm4j7tj2r8twwzwxs5z00000gn/T/ipykernel_67064/694894960.py:5: DtypeWarning: Columns (0: vFamilyTies, 1: jGeneNameTies, 2: jGeneAlleleTies) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  sample_df = pd.read_csv(filepath, sep='\\t', compression='gzip')\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "row0 = samplesheet.iloc[0]\n",
    "filepath = Path(row0['LOC']) / row0['SAMPLE']\n",
    "sample_short = row0['sample_short']\n",
    "\n",
    "sample_df = pd.read_csv(filepath, sep='\\t', compression='gzip')\n",
    "sample_df['sample_short'] = sample_short\n",
    "\n",
    "\n",
    "print(sample_df.head())\n",
    "print(sample_df.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "ef61e097",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "jGeneName                                      6\n",
       "jGeneAllele                                   11\n",
       "vMaxResolved                                  19\n",
       "vFamilyName                                   19\n",
       "vGeneName                                   2148\n",
       "aminoAcid                                   2527\n",
       "dMaxResolved                                3849\n",
       "dFamilyName                                 3849\n",
       "dGeneName                                   3849\n",
       "vGeneAllele                                 4620\n",
       "dGeneAllele                                 5917\n",
       "dFamilyTies                                12721\n",
       "dGeneNameTies                              12721\n",
       "vGeneAlleleTies                            13985\n",
       "vGeneNameTies                              14309\n",
       "dGeneAlleleTies                            14389\n",
       "vFamilyTies                                16438\n",
       "jGeneNameTies                              16451\n",
       "jGeneAlleleTies                            16452\n",
       "vAlignSubstitutionGeneThreePrimeIndexes    16457\n",
       "vOrphon                                    16457\n",
       "dOrphon                                    16457\n",
       "jOrphon                                    16457\n",
       "vFunction                                  16457\n",
       "dFunction                                  16457\n",
       "jFunction                                  16457\n",
       "fractionNucleated                          16457\n",
       "vAlignLength                               16457\n",
       "vAlignSubstitutionCount                    16457\n",
       "vAlignSubstitutionIndexes                  16457\n",
       "jFamilyTies                                16457\n",
       "vSeqWithMutations                          16457\n",
       "dtype: int64"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sumamrise NaN per column\n",
    "sample_df.isna().sum()[sample_df.isna().sum() > 0].sort_values(ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "66b785ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Grouping by jGeneName and vFamilyName being NA, and sequenceStatus (as fractions):\n",
      "       sequenceStatus\n",
      "False  In                83.114772\n",
      "       Out               15.292078\n",
      "       Stop               1.452260\n",
      "True   In                 0.081283\n",
      "       Out                0.054189\n",
      "       Stop               0.005419\n",
      "Name: count (templates/reads), dtype: float64\n",
      "Grouping by jGeneName/vGeneName being NA, and sequenceStatus (as fractions):\n",
      "       sequenceStatus\n",
      "False  In                72.282432\n",
      "       Out               13.601387\n",
      "       Stop               1.284274\n",
      "True   In                10.913623\n",
      "       Out                1.744879\n",
      "       Stop               0.173404\n",
      "Name: count (templates/reads), dtype: float64\n",
      "Collated Results:\n",
      "   level_0 sequenceStatus  Group1 Fraction  Group2 Fraction\n",
      "0    False             In        83.114772        72.282432\n",
      "1    False            Out        15.292078        13.601387\n",
      "2    False           Stop         1.452260         1.284274\n",
      "3     True             In         0.081283        10.913623\n",
      "4     True            Out         0.054189         1.744879\n",
      "5     True           Stop         0.005419         0.173404\n"
     ]
    }
   ],
   "source": [
    "# Check how many templates we will lose if we take only solved ones\n",
    "\n",
    "# Total sum of 'count (templates/reads)'\n",
    "total_count = sample_df['count (templates/reads)'].sum()\n",
    "\n",
    "# Group by whether jGeneName and vFamilyName are NA, and include sequenceStatus, compute the sum of 'count (templates/reads)'\n",
    "group1 = sample_df.groupby([sample_df[['jGeneName', 'vFamilyName']].isna().any(axis=1), 'sequenceStatus'])['count (templates/reads)'].sum()\n",
    "\n",
    "# Express as fractions of the total count\n",
    "group1_fraction = (group1 / total_count) * 100\n",
    "print(\"Grouping by jGeneName and vFamilyName being NA, and sequenceStatus (as fractions):\")\n",
    "print(group1_fraction)\n",
    "\n",
    "# Group by whether jGeneName is NA, include sequenceStatus, compute the sum of 'count (templates/reads)'\n",
    "group2 = sample_df.groupby([sample_df[['jGeneName','vGeneName']].isna().any(axis=1), 'sequenceStatus'])['count (templates/reads)'].sum()\n",
    "\n",
    "# Express as fractions of the total count\n",
    "group2_fraction = (group2 / total_count) * 100\n",
    "print(\"Grouping by jGeneName/vGeneName being NA, and sequenceStatus (as fractions):\")\n",
    "print(group2_fraction)\n",
    "\n",
    "# Collate the grouping results into a single DataFrame for better visualization\n",
    "collated_results = pd.concat([\n",
    "    group1_fraction.rename(\"Group1 Fraction\"),\n",
    "    group2_fraction.rename(\"Group2 Fraction\")\n",
    "], axis=1).reset_index()\n",
    "\n",
    "print(\"Collated Results:\")\n",
    "print(collated_results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "c0ff251d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sequences removed: 2\n",
      "Sum of counts removed: 3\n",
      "Productive sequences saved to: results_man/cleaned/F15625E_productive.tsv.gz\n",
      "Non-productive sequences saved to: results_man/cleaned/F15625E_nonproductive.tsv.gz\n",
      "Cleanup summary saved to: results_man/cleaned/F15625E_cleanup_summary.tsv\n"
     ]
    }
   ],
   "source": [
    "# Read in the file seqs_to_remove, which contains a list of CDR3 sequences to remove from the data\n",
    "seqs_to_remove_df = pd.read_csv(seqs_to_remove, sep='\\t', names=['nucleotide', 'aminoAcid'], header=None)\n",
    "\n",
    "# Remove sequences by matching their nucleotide and aminoAcid\n",
    "merged_df = sample_df.merge(seqs_to_remove_df, on=['nucleotide', 'aminoAcid'], how='left', indicator=True)\n",
    "removed_sequences = merged_df[merged_df['_merge'] == 'both']\n",
    "remaining_sequences = merged_df[merged_df['_merge'] == 'left_only']\n",
    "\n",
    "# Write info about how many sequences were found and removed\n",
    "num_removed_rows = len(removed_sequences)\n",
    "removed_counts_sum = removed_sequences['count (templates/reads)'].sum()\n",
    "print(f\"Number of sequences removed: {num_removed_rows}\")\n",
    "print(f\"Sum of counts removed: {removed_counts_sum}\")\n",
    "\n",
    "# Take only rows where jGeneName and vFamilyName are not NA\n",
    "filtered_df = remaining_sequences[~remaining_sequences[['jGeneName', 'vFamilyName']].isna().any(axis=1)]\n",
    "\n",
    "# Split into productive and non-productive sequences\n",
    "productive_df = filtered_df[filtered_df['sequenceStatus'] == 'In']\n",
    "nonproductive_df = filtered_df[filtered_df['sequenceStatus'] != 'In']\n",
    "\n",
    "# Save the outputs\n",
    "productive_output_path = f\"{output_loc}/{sample_short}_productive.tsv.gz\"\n",
    "nonproductive_output_path = f\"{output_loc}/{sample_short}_nonproductive.tsv.gz\"\n",
    "productive_df.to_csv(productive_output_path, sep='\\t', index=False, compression='gzip')\n",
    "nonproductive_df.to_csv(nonproductive_output_path, sep='\\t', index=False, compression='gzip')\n",
    "\n",
    "print(f\"Productive sequences saved to: {productive_output_path}\")\n",
    "print(f\"Non-productive sequences saved to: {nonproductive_output_path}\")\n",
    "\n",
    "# Create a cleanup summary DataFrame\n",
    "cleanup_summary = pd.DataFrame({\n",
    "    'Category': ['Removed Sequences', 'Productive Sequences', 'Non-Productive Sequences'],\n",
    "    'Num Sequences': [num_removed_rows, len(productive_df), len(nonproductive_df)],\n",
    "    'Sum of Counts': [removed_counts_sum, productive_df['count (templates/reads)'].sum(), nonproductive_df['count (templates/reads)'].sum()]\n",
    "})\n",
    "\n",
    "# Save the cleanup summary\n",
    "cleanup_summary_path = f\"{output_loc}/{sample_short}_cleanup_summary.tsv\"\n",
    "cleanup_summary.to_csv(cleanup_summary_path, sep='\\t', index=False)\n",
    "print(f\"Cleanup summary saved to: {cleanup_summary_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12f43226",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "e0053b54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                          nucleotide          aminoAcid  \\\n",
      "0  CACCTACACACCCTGCAGCCAGAAGACTCGGCCCTGTATCTCTGCG...      CASSQDKNQPQHF   \n",
      "1  ATCCAGCGCACACAGCAGGAGGACTCGGCCGTGTATCTCTGTGCCA...     CASSLMSGPGELFF   \n",
      "2  TCAGAACCCAGGGACTCAGCTGTGTACTTCTGTGCCAGCAGACGCC...  CASRRPTGQGIRSGYTF   \n",
      "3  ATCAATTCCCTGGAGCTTGGTGACTCTGCTGTGTATTTCTGTGCCA...     CASSQGTGGSPLHF   \n",
      "4  GATCCAGCGCACAGAGCAGGGGGACTCAGCTGTGTATCTCTGTGCC...                NaN   \n",
      "\n",
      "   count (templates/reads)  frequencyCount (%)  cdr3Length  \\\n",
      "0                       14            0.000759          39   \n",
      "1                        5            0.000271          42   \n",
      "2                        9            0.000488          51   \n",
      "3                        8            0.000434          42   \n",
      "4                        4            0.000217          41   \n",
      "\n",
      "          vMaxResolved vFamilyName         vGeneName  vGeneAllele vFamilyTies  \\\n",
      "0        TCRBV04-03*01     TCRBV04        TCRBV04-03          1.0         NaN   \n",
      "1           TCRBV07-02     TCRBV07        TCRBV07-02          NaN         NaN   \n",
      "2              TCRBV12     TCRBV12               NaN          NaN         NaN   \n",
      "3  TCRBV03-01/03-02*01     TCRBV03  TCRBV03-01/03-02          1.0         NaN   \n",
      "4        TCRBV07-04*01     TCRBV07        TCRBV07-04          1.0         NaN   \n",
      "\n",
      "                 vGeneNameTies vGeneAlleleTies   dMaxResolved dFamilyName  \\\n",
      "0                          NaN             NaN            NaN         NaN   \n",
      "1                          NaN           01,02     TCRBD02-01     TCRBD02   \n",
      "2  TCRBV12-03/12-04,TCRBV12-04             NaN  TCRBD01-01*01     TCRBD01   \n",
      "3                          NaN             NaN  TCRBD01-01*01     TCRBD01   \n",
      "4                          NaN             NaN  TCRBD01-01*01     TCRBD01   \n",
      "\n",
      "    dGeneName  dGeneAllele      dFamilyTies          dGeneNameTies  \\\n",
      "0         NaN          NaN  TCRBD01,TCRBD02  TCRBD01-01,TCRBD02-01   \n",
      "1  TCRBD02-01          NaN              NaN                    NaN   \n",
      "2  TCRBD01-01          1.0              NaN                    NaN   \n",
      "3  TCRBD01-01          1.0              NaN                    NaN   \n",
      "4  TCRBD01-01          1.0              NaN                    NaN   \n",
      "\n",
      "  dGeneAlleleTies   jMaxResolved jFamilyName   jGeneName  jGeneAllele  \\\n",
      "0             NaN  TCRBJ01-05*01     TCRBJ01  TCRBJ01-05          1.0   \n",
      "1           01,02  TCRBJ02-02*01     TCRBJ02  TCRBJ02-02          1.0   \n",
      "2             NaN  TCRBJ01-02*01     TCRBJ01  TCRBJ01-02          1.0   \n",
      "3             NaN  TCRBJ01-06*02     TCRBJ01  TCRBJ01-06          2.0   \n",
      "4             NaN  TCRBJ01-02*01     TCRBJ01  TCRBJ01-02          1.0   \n",
      "\n",
      "   jFamilyTies jGeneNameTies jGeneAlleleTies  vDeletion  n1Insertion  \\\n",
      "0          NaN           NaN             NaN          0            2   \n",
      "1          NaN           NaN             NaN          2            4   \n",
      "2          NaN           NaN             NaN          6            9   \n",
      "3          NaN           NaN             NaN          1            2   \n",
      "4          NaN           NaN             NaN          4            3   \n",
      "\n",
      "   d5Deletion  d3Deletion  n2Insertion  jDeletion  vIndex  n1Index  dIndex  \\\n",
      "0           5           5            0          4      42       59      61   \n",
      "1           8           4            2          6      39       54      58   \n",
      "2           0           6           13          8      30       41      50   \n",
      "3           3           1            0          9      39       55      57   \n",
      "4           2           5            3          3      40       53      56   \n",
      "\n",
      "   n2Index  jIndex  estimatedNumberGenomes sequenceStatus cloneResolved  \\\n",
      "0       -1      63                      14             In           VDJ   \n",
      "1       62      64                       5             In           VDJ   \n",
      "2       56      69                       9             In           VDJ   \n",
      "3       -1      65                       8             In           VDJ   \n",
      "4       61      64                       4            Out           VDJ   \n",
      "\n",
      "   vOrphon  dOrphon  jOrphon  vFunction  dFunction  jFunction  \\\n",
      "0      NaN      NaN      NaN        NaN        NaN        NaN   \n",
      "1      NaN      NaN      NaN        NaN        NaN        NaN   \n",
      "2      NaN      NaN      NaN        NaN        NaN        NaN   \n",
      "3      NaN      NaN      NaN        NaN        NaN        NaN   \n",
      "4      NaN      NaN      NaN        NaN        NaN        NaN   \n",
      "\n",
      "   fractionNucleated  vAlignLength  vAlignSubstitutionCount  \\\n",
      "0                NaN           NaN                      NaN   \n",
      "1                NaN           NaN                      NaN   \n",
      "2                NaN           NaN                      NaN   \n",
      "3                NaN           NaN                      NaN   \n",
      "4                NaN           NaN                      NaN   \n",
      "\n",
      "   vAlignSubstitutionIndexes  vAlignSubstitutionGeneThreePrimeIndexes  \\\n",
      "0                        NaN                                      NaN   \n",
      "1                        NaN                                      NaN   \n",
      "2                        NaN                                      NaN   \n",
      "3                        NaN                                      NaN   \n",
      "4                        NaN                                      NaN   \n",
      "\n",
      "   vSeqWithMutations sample_short  \n",
      "0                NaN      F15625E  \n",
      "1                NaN      F15625E  \n",
      "2                NaN      F15625E  \n",
      "3                NaN      F15625E  \n",
      "4                NaN      F15625E  \n"
     ]
    }
   ],
   "source": [
    "# Print the first 5 rows of sample_df with all columns\n",
    "pd.set_option('display.max_columns', None)  # Ensure all columns are displayed\n",
    "print(sample_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "3c610f59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summary of 'sequenceStatus':\n",
      "sequenceStatus\n",
      "In      13688\n",
      "Out      2527\n",
      "Stop      242\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Summarise the content of the column 'sequenceStatus'\n",
    "sequence_status_summary = sample_df['sequenceStatus'].value_counts()\n",
    "print(\"Summary of 'sequenceStatus':\")\n",
    "print(sequence_status_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "dd72ccae",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Missing required columns: ['repertoire_id', 'junction_aa', 'duplicate_count']",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[94]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# Calculate repertoire diversity metrics\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m diversity = \u001b[43mls\u001b[49m\u001b[43m.\u001b[49m\u001b[43mclonality\u001b[49m\u001b[43m(\u001b[49m\u001b[43msample_df\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m      4\u001b[39m \u001b[38;5;66;03m# Visualize clonal expansion\u001b[39;00m\n\u001b[32m      5\u001b[39m fig = ls.plot_clonality(sample_df)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/lib/python3.12/site-packages/lymphoseq/analysis/diversity.py:61\u001b[39m, in \u001b[36mclonality\u001b[39m\u001b[34m(data, rarefy, min_count, iterations, group_by)\u001b[39m\n\u001b[32m     59\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m _clonality_polars(data, rarefy, min_count, iterations, group_by)\n\u001b[32m     60\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m---> \u001b[39m\u001b[32m61\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_clonality_pandas\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrarefy\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmin_count\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43miterations\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgroup_by\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/lib/python3.12/site-packages/lymphoseq/analysis/diversity.py:146\u001b[39m, in \u001b[36m_clonality_pandas\u001b[39m\u001b[34m(data, rarefy, min_count, iterations, group_by)\u001b[39m\n\u001b[32m    144\u001b[39m missing_cols = [col \u001b[38;5;28;01mfor\u001b[39;00m col \u001b[38;5;129;01min\u001b[39;00m required_cols \u001b[38;5;28;01mif\u001b[39;00m col \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m data.columns]\n\u001b[32m    145\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m missing_cols:\n\u001b[32m--> \u001b[39m\u001b[32m146\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mMissing required columns: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmissing_cols\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m)\n\u001b[32m    148\u001b[39m \u001b[38;5;66;03m# Filter productive sequences\u001b[39;00m\n\u001b[32m    149\u001b[39m productive_data = data[\n\u001b[32m    150\u001b[39m     (data[\u001b[33m\"\u001b[39m\u001b[33mproductive\u001b[39m\u001b[33m\"\u001b[39m] == \u001b[38;5;28;01mTrue\u001b[39;00m) &\n\u001b[32m    151\u001b[39m     (data[\u001b[33m\"\u001b[39m\u001b[33mjunction_aa\u001b[39m\u001b[33m\"\u001b[39m] != \u001b[33m\"\u001b[39m\u001b[33m\"\u001b[39m) &\n\u001b[32m    152\u001b[39m     (data[\u001b[33m\"\u001b[39m\u001b[33mjunction_aa\u001b[39m\u001b[33m\"\u001b[39m].notna())\n\u001b[32m    153\u001b[39m ].copy()\n",
      "\u001b[31mValueError\u001b[39m: Missing required columns: ['repertoire_id', 'junction_aa', 'duplicate_count']"
     ]
    }
   ],
   "source": [
    "# Calculate repertoire diversity metrics\n",
    "diversity = ls.clonality(sample_df)\n",
    "\n",
    "# Visualize clonal expansion\n",
    "fig = ls.plot_clonality(sample_df)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69b3b794",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing files: 100%|██████████| 1/1 [00:00<00:00, 12.92file/s, records=16,457, current=16,457]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Completed: 1 files processed, 16,457 total records standardized\n",
      "['sequence_id', 'repertoire_id', 'sequence', 'sequence_aa', 'junction', 'junction_aa', 'junction_length', 'junction_aa_length', 'v_call', 'd_call', 'j_call', 'c_call', 'productive', 'vj_in_frame', 'stop_codon', 'locus', 'duplicate_count', 'duplicate_frequency', 'consensus_count', 'v_family', 'd_family', 'j_family', 'cdr1', 'cdr1_aa', 'cdr2', 'cdr2_aa', 'cdr3', 'cdr3_aa', 'clone_id', 'cell_id', 'nucleotide', 'aminoAcid', 'count (templates/reads)', 'frequencyCount (%)', 'cdr3Length', 'vMaxResolved', 'vFamilyName', 'vGeneName', 'vGeneAllele', 'vFamilyTies', 'vGeneNameTies', 'vGeneAlleleTies', 'dMaxResolved', 'dFamilyName', 'dGeneName', 'dGeneAllele', 'dFamilyTies', 'dGeneNameTies', 'dGeneAlleleTies', 'jMaxResolved', 'jFamilyName', 'jGeneName', 'jGeneAllele', 'jFamilyTies', 'jGeneNameTies', 'jGeneAlleleTies', 'v_deletions', 'n1Insertion', 'd5_deletions', 'd3_deletions', 'n2Insertion', 'j_deletions', 'vIndex', 'n1Index', 'dIndex', 'n2Index', 'jIndex', 'sequenceStatus', 'cloneResolved', 'vOrphon', 'dOrphon', 'jOrphon', 'vFunction', 'dFunction', 'jFunction', 'fractionNucleated', 'vAlignLength', 'vAlignSubstitutionCount', 'vAlignSubstitutionIndexes', 'vAlignSubstitutionGeneThreePrimeIndexes', 'vSeqWithMutations']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "data = ls.read_immunoseq(filepath )\n",
    "print(data.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84273b1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def immunoseq_to_airr(df: pl.DataFrame) -> pl.DataFrame:\n",
    "    \"\"\"\n",
    "    Rename ImmunoSEQ columns to AIRR Rearrangement standard column names and add a productive column.\n",
    "\n",
    "    Args:\n",
    "        df: Polars DataFrame with ImmunoSEQ-style columns.\n",
    "\n",
    "    Returns:\n",
    "        A new Polars DataFrame with AIRR standard column names and a productive column.\n",
    "    \"\"\"\n",
    "\n",
    "    # Define the column mapping\n",
    "    col_map = {\n",
    "        \"nucleotide\": \"sequence\",\n",
    "        \"aminoAcid\": \"junction_aa\",\n",
    "        'cdr3Length':'junction_aa_length',\n",
    "        \"vGeneName\": \"v_call\",\n",
    "        \"jGeneName\": \"j_call\",\n",
    "        \"sequenceStatus\": \"frame_type\",\n",
    "        \"count (templates/reads)\": \"duplicate_count\",\n",
    "        \"vFamilyName\": \"v_family\",\n",
    "        \"jFamilyName\": \"j_family\",\n",
    "        \"vGeneAllele\": \"v_allele\",\n",
    "        \"jGeneAllele\": \"j_allele\",\n",
    "        \"dGeneName\": \"d_gene\"\n",
    "    }\n",
    "\n",
    "    # Only keep mappings where the column exists in df\n",
    "    existing_map = {old: new for old, new in col_map.items() if old in df.columns}\n",
    "    print(f\"Existing columns to rename: {list(existing_map.keys())}\")\n",
    "\n",
    "    target_cols = list(existing_map.values())\n",
    "    cols_to_drop = [col for col in target_cols if col in df.columns]\n",
    "    cols_to_drop += [\"sequence_aa\",\"junction\",\"junction_length\",\"d_call\",\"c_call\",\"productive\",\"vj_in_frame\",\"stop_codon\",\"locus\",\"duplicate_frequency\",\"consensus_count\",\"d_family\",\"cdr1\",\"cdr1_aa\",\"cdr2\",\"cdr2_aa\",\"cdr3\",\"cdr3_aa\",\"clone_id\",\"cell_id\"]\n",
    "    \n",
    "    if cols_to_drop:\n",
    "        print(f\"Dropping existing columns to avoid conflicts: {cols_to_drop}\")\n",
    "        df = df.drop(cols_to_drop)\n",
    "\n",
    "    # Perform renaming\n",
    "    df = df.rename(existing_map)\n",
    "\n",
    "   # Cast columns with 'length' or 'count' in the name to integers\n",
    "    int_cols = [col for col in df.columns if \"length\" in col.lower() or \"count\" in col.lower()]\n",
    "    if int_cols:\n",
    "        print(f\"Casting columns to integers: {int_cols}\")\n",
    "        df = df.with_columns([pl.col(col).cast(pl.Int64, strict=False) for col in int_cols])\n",
    "\n",
    "    # Add a productive column based on frame_type\n",
    "    if \"frame_type\" in df.columns:\n",
    "        df = df.with_columns(\n",
    "            (pl.col(\"frame_type\") == \"In\").alias(\"productive\")\n",
    "        )\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e5e9910",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Existing columns to rename: ['nucleotide', 'aminoAcid', 'cdr3Length', 'vGeneName', 'jGeneName', 'sequenceStatus', 'count (templates/reads)', 'vFamilyName', 'jFamilyName', 'vGeneAllele', 'jGeneAllele', 'dGeneName']\n",
      "Dropping existing columns to avoid conflicts: ['sequence', 'junction_aa', 'junction_aa_length', 'v_call', 'j_call', 'duplicate_count', 'v_family', 'j_family', 'sequence_aa', 'junction', 'junction_length', 'd_call', 'c_call', 'productive', 'vj_in_frame', 'stop_codon', 'locus', 'duplicate_frequency', 'consensus_count', 'd_family', 'cdr1', 'cdr1_aa', 'cdr2', 'cdr2_aa', 'cdr3', 'cdr3_aa', 'clone_id', 'cell_id']\n",
      "Casting columns to integers: ['duplicate_count', 'frequencyCount (%)', 'junction_aa_length', 'vAlignLength', 'vAlignSubstitutionCount']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (16_457, 55)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>sequence_id</th><th>repertoire_id</th><th>sequence</th><th>junction_aa</th><th>duplicate_count</th><th>frequencyCount (%)</th><th>junction_aa_length</th><th>vMaxResolved</th><th>v_family</th><th>v_call</th><th>v_allele</th><th>vFamilyTies</th><th>vGeneNameTies</th><th>vGeneAlleleTies</th><th>dMaxResolved</th><th>dFamilyName</th><th>d_gene</th><th>dGeneAllele</th><th>dFamilyTies</th><th>dGeneNameTies</th><th>dGeneAlleleTies</th><th>jMaxResolved</th><th>j_family</th><th>j_call</th><th>j_allele</th><th>jFamilyTies</th><th>jGeneNameTies</th><th>jGeneAlleleTies</th><th>v_deletions</th><th>n1Insertion</th><th>d5_deletions</th><th>d3_deletions</th><th>n2Insertion</th><th>j_deletions</th><th>vIndex</th><th>n1Index</th><th>dIndex</th><th>n2Index</th><th>jIndex</th><th>frame_type</th><th>cloneResolved</th><th>vOrphon</th><th>dOrphon</th><th>jOrphon</th><th>vFunction</th><th>dFunction</th><th>jFunction</th><th>fractionNucleated</th><th>vAlignLength</th><th>vAlignSubstitutionCount</th><th>vAlignSubstitutionIndexes</th><th>vAlignSubstitutionGeneThreePrimeIndexes</th><th>vSeqWithMutations</th><th>productive</th><th>shannon_diversity</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>bool</td><td>f64</td></tr></thead><tbody><tr><td>&quot;&quot;</td><td>&quot;FIN_15625_AgXP_TCRB&quot;</td><td>&quot;CACCTACACACCCTGCAGCCAGAAGACTCG…</td><td>&quot;CASSQDKNQPQHF&quot;</td><td>14</td><td>null</td><td>39</td><td>&quot;TCRBV04-03*01&quot;</td><td>&quot;TCRBV04&quot;</td><td>&quot;TCRBV04-03&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBD01,TCRBD02&quot;</td><td>&quot;TCRBD01-01,TCRBD02-01&quot;</td><td>null</td><td>&quot;TCRBJ01-05*01&quot;</td><td>&quot;TCRBJ01&quot;</td><td>&quot;TCRBJ01-05&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;0&quot;</td><td>&quot;2&quot;</td><td>&quot;5&quot;</td><td>&quot;5&quot;</td><td>&quot;0&quot;</td><td>&quot;4&quot;</td><td>&quot;42&quot;</td><td>&quot;59&quot;</td><td>&quot;61&quot;</td><td>&quot;-1&quot;</td><td>&quot;63&quot;</td><td>&quot;In&quot;</td><td>&quot;VDJ&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>9.458566</td></tr><tr><td>&quot;&quot;</td><td>&quot;FIN_15625_AgXP_TCRB&quot;</td><td>&quot;ATCCAGCGCACACAGCAGGAGGACTCGGCC…</td><td>&quot;CASSLMSGPGELFF&quot;</td><td>5</td><td>null</td><td>42</td><td>&quot;TCRBV07-02&quot;</td><td>&quot;TCRBV07&quot;</td><td>&quot;TCRBV07-02&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;01,02&quot;</td><td>&quot;TCRBD02-01&quot;</td><td>&quot;TCRBD02&quot;</td><td>&quot;TCRBD02-01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;01,02&quot;</td><td>&quot;TCRBJ02-02*01&quot;</td><td>&quot;TCRBJ02&quot;</td><td>&quot;TCRBJ02-02&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;2&quot;</td><td>&quot;4&quot;</td><td>&quot;8&quot;</td><td>&quot;4&quot;</td><td>&quot;2&quot;</td><td>&quot;6&quot;</td><td>&quot;39&quot;</td><td>&quot;54&quot;</td><td>&quot;58&quot;</td><td>&quot;62&quot;</td><td>&quot;64&quot;</td><td>&quot;In&quot;</td><td>&quot;VDJ&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>9.458566</td></tr><tr><td>&quot;&quot;</td><td>&quot;FIN_15625_AgXP_TCRB&quot;</td><td>&quot;TCAGAACCCAGGGACTCAGCTGTGTACTTC…</td><td>&quot;CASRRPTGQGIRSGYTF&quot;</td><td>9</td><td>null</td><td>51</td><td>&quot;TCRBV12&quot;</td><td>&quot;TCRBV12&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBV12-03/12-04,TCRBV12-04&quot;</td><td>null</td><td>&quot;TCRBD01-01*01&quot;</td><td>&quot;TCRBD01&quot;</td><td>&quot;TCRBD01-01&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBJ01-02*01&quot;</td><td>&quot;TCRBJ01&quot;</td><td>&quot;TCRBJ01-02&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;6&quot;</td><td>&quot;9&quot;</td><td>&quot;0&quot;</td><td>&quot;6&quot;</td><td>&quot;13&quot;</td><td>&quot;8&quot;</td><td>&quot;30&quot;</td><td>&quot;41&quot;</td><td>&quot;50&quot;</td><td>&quot;56&quot;</td><td>&quot;69&quot;</td><td>&quot;In&quot;</td><td>&quot;VDJ&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>9.458566</td></tr><tr><td>&quot;&quot;</td><td>&quot;FIN_15625_AgXP_TCRB&quot;</td><td>&quot;ATCAATTCCCTGGAGCTTGGTGACTCTGCT…</td><td>&quot;CASSQGTGGSPLHF&quot;</td><td>8</td><td>null</td><td>42</td><td>&quot;TCRBV03-01/03-02*01&quot;</td><td>&quot;TCRBV03&quot;</td><td>&quot;TCRBV03-01/03-02&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBD01-01*01&quot;</td><td>&quot;TCRBD01&quot;</td><td>&quot;TCRBD01-01&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBJ01-06*02&quot;</td><td>&quot;TCRBJ01&quot;</td><td>&quot;TCRBJ01-06&quot;</td><td>&quot;02&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;1&quot;</td><td>&quot;2&quot;</td><td>&quot;3&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td><td>&quot;9&quot;</td><td>&quot;39&quot;</td><td>&quot;55&quot;</td><td>&quot;57&quot;</td><td>&quot;-1&quot;</td><td>&quot;65&quot;</td><td>&quot;In&quot;</td><td>&quot;VDJ&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>9.458566</td></tr><tr><td>&quot;&quot;</td><td>&quot;FIN_15625_AgXP_TCRB&quot;</td><td>&quot;GATCCAGCGCACAGAGCAGGGGGACTCAGC…</td><td>null</td><td>4</td><td>null</td><td>41</td><td>&quot;TCRBV07-04*01&quot;</td><td>&quot;TCRBV07&quot;</td><td>&quot;TCRBV07-04&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBD01-01*01&quot;</td><td>&quot;TCRBD01&quot;</td><td>&quot;TCRBD01-01&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBJ01-02*01&quot;</td><td>&quot;TCRBJ01&quot;</td><td>&quot;TCRBJ01-02&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;4&quot;</td><td>&quot;3&quot;</td><td>&quot;2&quot;</td><td>&quot;5&quot;</td><td>&quot;3&quot;</td><td>&quot;3&quot;</td><td>&quot;40&quot;</td><td>&quot;53&quot;</td><td>&quot;56&quot;</td><td>&quot;61&quot;</td><td>&quot;64&quot;</td><td>&quot;Out&quot;</td><td>&quot;VDJ&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>false</td><td>9.458566</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;&quot;</td><td>&quot;FIN_15625_AgXP_TCRB&quot;</td><td>&quot;CATCAATTCCCTGGAGCTTGGTGACTCTGC…</td><td>null</td><td>1</td><td>null</td><td>41</td><td>&quot;TCRBV03-01/03-02*01&quot;</td><td>&quot;TCRBV03&quot;</td><td>&quot;TCRBV03-01/03-02&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBD02-01&quot;</td><td>&quot;TCRBD02&quot;</td><td>&quot;TCRBD02-01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;01,02&quot;</td><td>&quot;TCRBJ02-02*01&quot;</td><td>&quot;TCRBJ02&quot;</td><td>&quot;TCRBJ02-02&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;7&quot;</td><td>&quot;6&quot;</td><td>&quot;8&quot;</td><td>&quot;3&quot;</td><td>&quot;15&quot;</td><td>&quot;18&quot;</td><td>&quot;40&quot;</td><td>&quot;50&quot;</td><td>&quot;56&quot;</td><td>&quot;61&quot;</td><td>&quot;76&quot;</td><td>&quot;Out&quot;</td><td>&quot;VDJ&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>false</td><td>9.458566</td></tr><tr><td>&quot;&quot;</td><td>&quot;FIN_15625_AgXP_TCRB&quot;</td><td>&quot;AAGATCCAGCCTGCAGAGCTTGGGGACTCG…</td><td>&quot;CASSLDPDSPAFF&quot;</td><td>1</td><td>null</td><td>39</td><td>&quot;TCRBV11-03*01&quot;</td><td>&quot;TCRBV11&quot;</td><td>&quot;TCRBV11-03&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBD02-01&quot;</td><td>&quot;TCRBD02&quot;</td><td>&quot;TCRBD02-01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;01,02&quot;</td><td>&quot;TCRBJ01-01*01&quot;</td><td>&quot;TCRBJ01&quot;</td><td>&quot;TCRBJ01-01&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;0&quot;</td><td>&quot;4&quot;</td><td>&quot;2&quot;</td><td>&quot;10&quot;</td><td>&quot;4&quot;</td><td>&quot;10&quot;</td><td>&quot;42&quot;</td><td>&quot;59&quot;</td><td>&quot;63&quot;</td><td>&quot;67&quot;</td><td>&quot;71&quot;</td><td>&quot;In&quot;</td><td>&quot;VDJ&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>9.458566</td></tr><tr><td>&quot;&quot;</td><td>&quot;FIN_15625_AgXP_TCRB&quot;</td><td>&quot;CACATCAATTCCCTGGAGCTTGGTGACTCT…</td><td>&quot;CASSLGGATEAFF&quot;</td><td>1</td><td>null</td><td>39</td><td>&quot;TCRBV03-01/03-02*01&quot;</td><td>&quot;TCRBV03&quot;</td><td>&quot;TCRBV03-01/03-02&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBD02-01*01&quot;</td><td>&quot;TCRBD02&quot;</td><td>&quot;TCRBD02-01&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBJ01-01*01&quot;</td><td>&quot;TCRBJ01&quot;</td><td>&quot;TCRBJ01-01&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;4&quot;</td><td>&quot;1&quot;</td><td>&quot;8&quot;</td><td>&quot;0&quot;</td><td>&quot;1&quot;</td><td>&quot;4&quot;</td><td>&quot;42&quot;</td><td>&quot;55&quot;</td><td>&quot;56&quot;</td><td>&quot;64&quot;</td><td>&quot;65&quot;</td><td>&quot;In&quot;</td><td>&quot;VDJ&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>9.458566</td></tr><tr><td>&quot;&quot;</td><td>&quot;FIN_15625_AgXP_TCRB&quot;</td><td>&quot;AGTGACCAGTGCCCATCCTGAAGACAGCAG…</td><td>null</td><td>1</td><td>null</td><td>41</td><td>&quot;TCRBV20&quot;</td><td>&quot;TCRBV20&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBV20-01,TCRBV20-or09_02&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBD01,TCRBD02&quot;</td><td>&quot;TCRBD01-01,TCRBD02-01&quot;</td><td>null</td><td>&quot;TCRBJ01-01*01&quot;</td><td>&quot;TCRBJ01&quot;</td><td>&quot;TCRBJ01-01&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;7&quot;</td><td>&quot;5&quot;</td><td>&quot;5&quot;</td><td>&quot;2&quot;</td><td>&quot;8&quot;</td><td>&quot;4&quot;</td><td>&quot;40&quot;</td><td>&quot;47&quot;</td><td>&quot;52&quot;</td><td>&quot;57&quot;</td><td>&quot;65&quot;</td><td>&quot;Out&quot;</td><td>&quot;VDJ&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>false</td><td>9.458566</td></tr><tr><td>&quot;&quot;</td><td>&quot;FIN_15625_AgXP_TCRB&quot;</td><td>&quot;GGAGTCAGGGGACACAGCACTGTATTTCTG…</td><td>null</td><td>1</td><td>null</td><td>53</td><td>&quot;TCRBV21-01*01&quot;</td><td>&quot;TCRBV21&quot;</td><td>&quot;TCRBV21-01&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBD01-01*01&quot;</td><td>&quot;TCRBD01&quot;</td><td>&quot;TCRBD01-01&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;TCRBJ01-01*01&quot;</td><td>&quot;TCRBJ01&quot;</td><td>&quot;TCRBJ01-01&quot;</td><td>&quot;01&quot;</td><td>null</td><td>null</td><td>null</td><td>&quot;3&quot;</td><td>&quot;6&quot;</td><td>&quot;2&quot;</td><td>&quot;0&quot;</td><td>&quot;4&quot;</td><td>&quot;1&quot;</td><td>&quot;28&quot;</td><td>&quot;42&quot;</td><td>&quot;48&quot;</td><td>&quot;58&quot;</td><td>&quot;62&quot;</td><td>&quot;Out&quot;</td><td>&quot;VDJ&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>false</td><td>9.458566</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (16_457, 55)\n",
       "┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐\n",
       "│ sequence_ ┆ repertoir ┆ sequence  ┆ junction_ ┆ … ┆ vAlignSub ┆ vSeqWithM ┆ productiv ┆ shannon_ │\n",
       "│ id        ┆ e_id      ┆ ---       ┆ aa        ┆   ┆ stitution ┆ utations  ┆ e         ┆ diversit │\n",
       "│ ---       ┆ ---       ┆ str       ┆ ---       ┆   ┆ GeneThree ┆ ---       ┆ ---       ┆ y        │\n",
       "│ str       ┆ str       ┆           ┆ str       ┆   ┆ Pri…      ┆ str       ┆ bool      ┆ ---      │\n",
       "│           ┆           ┆           ┆           ┆   ┆ ---       ┆           ┆           ┆ f64      │\n",
       "│           ┆           ┆           ┆           ┆   ┆ str       ┆           ┆           ┆          │\n",
       "╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡\n",
       "│           ┆ FIN_15625 ┆ CACCTACAC ┆ CASSQDKNQ ┆ … ┆ null      ┆ null      ┆ true      ┆ 9.458566 │\n",
       "│           ┆ _AgXP_TCR ┆ ACCCTGCAG ┆ PQHF      ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ B         ┆ CCAGAAGAC ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆           ┆ TCG…      ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ FIN_15625 ┆ ATCCAGCGC ┆ CASSLMSGP ┆ … ┆ null      ┆ null      ┆ true      ┆ 9.458566 │\n",
       "│           ┆ _AgXP_TCR ┆ ACACAGCAG ┆ GELFF     ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ B         ┆ GAGGACTCG ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆           ┆ GCC…      ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ FIN_15625 ┆ TCAGAACCC ┆ CASRRPTGQ ┆ … ┆ null      ┆ null      ┆ true      ┆ 9.458566 │\n",
       "│           ┆ _AgXP_TCR ┆ AGGGACTCA ┆ GIRSGYTF  ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ B         ┆ GCTGTGTAC ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆           ┆ TTC…      ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ FIN_15625 ┆ ATCAATTCC ┆ CASSQGTGG ┆ … ┆ null      ┆ null      ┆ true      ┆ 9.458566 │\n",
       "│           ┆ _AgXP_TCR ┆ CTGGAGCTT ┆ SPLHF     ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ B         ┆ GGTGACTCT ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆           ┆ GCT…      ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ FIN_15625 ┆ GATCCAGCG ┆ null      ┆ … ┆ null      ┆ null      ┆ false     ┆ 9.458566 │\n",
       "│           ┆ _AgXP_TCR ┆ CACAGAGCA ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ B         ┆ GGGGGACTC ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆           ┆ AGC…      ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …        │\n",
       "│           ┆ FIN_15625 ┆ CATCAATTC ┆ null      ┆ … ┆ null      ┆ null      ┆ false     ┆ 9.458566 │\n",
       "│           ┆ _AgXP_TCR ┆ CCTGGAGCT ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ B         ┆ TGGTGACTC ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆           ┆ TGC…      ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ FIN_15625 ┆ AAGATCCAG ┆ CASSLDPDS ┆ … ┆ null      ┆ null      ┆ true      ┆ 9.458566 │\n",
       "│           ┆ _AgXP_TCR ┆ CCTGCAGAG ┆ PAFF      ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ B         ┆ CTTGGGGAC ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆           ┆ TCG…      ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ FIN_15625 ┆ CACATCAAT ┆ CASSLGGAT ┆ … ┆ null      ┆ null      ┆ true      ┆ 9.458566 │\n",
       "│           ┆ _AgXP_TCR ┆ TCCCTGGAG ┆ EAFF      ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ B         ┆ CTTGGTGAC ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆           ┆ TCT…      ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ FIN_15625 ┆ AGTGACCAG ┆ null      ┆ … ┆ null      ┆ null      ┆ false     ┆ 9.458566 │\n",
       "│           ┆ _AgXP_TCR ┆ TGCCCATCC ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ B         ┆ TGAAGACAG ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆           ┆ CAG…      ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ FIN_15625 ┆ GGAGTCAGG ┆ null      ┆ … ┆ null      ┆ null      ┆ false     ┆ 9.458566 │\n",
       "│           ┆ _AgXP_TCR ┆ GGACACAGC ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ B         ┆ ACTGTATTT ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆           ┆ CTG…      ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "airr_df = immunoseq_to_airr(data)\n",
    "airr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f63bf0be",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "326b10fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to calculate Shannon diversity and return it as a single number\n",
    "def compute_diversity_metrics(df: pl.DataFrame) -> pl.DataFrame:\n",
    "    def shannon_diversity(df: pl.DataFrame, count_col=\"duplicate_count\") -> float:\n",
    "        if count_col in df.columns:\n",
    "            total_count = df[count_col].sum()\n",
    "            shannon_div = (\n",
    "                -df\n",
    "                .with_columns((pl.col(count_col) / total_count).alias(\"proportion\"))\n",
    "                .with_columns((pl.col(\"proportion\") * pl.col(\"proportion\").log()).alias(\"shannon_term\"))\n",
    "                [\"shannon_term\"]\n",
    "                .sum()\n",
    "            )\n",
    "            return shannon_div\n",
    "        return 0.0\n",
    "\n",
    "    # Calculate Shannon diversity as a single number\n",
    "    shannon_div = shannon_diversity(airr_df)\n",
    "\n",
    "    # Calculate other diversity metrics\n",
    "    diversity = ls.diversity_metrics(airr_df)\n",
    "    isinstance(diversity, pl.DataFrame)\n",
    "\n",
    "    # Add shannon_div as a new column to diversity (handles polars DataFrame, pandas DataFrame/Series, or scalars)\n",
    "    diversity = diversity.with_columns(pl.lit(shannon_div).alias(\"shannon_diversity\"))\n",
    "    return diversity\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "511f302c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1, 9)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>repertoire_id</th><th>total_sequences</th><th>unique_productive_sequences</th><th>total_count</th><th>clonality</th><th>gini_coefficient</th><th>top_productive_sequence</th><th>convergence</th><th>shannon_diversity</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>i64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;FIN_15625_AgXP_TCRB&quot;</td><td>13688</td><td>13541</td><td>15353</td><td>0.005772</td><td>0.101629</td><td>0.001238</td><td>1.010856</td><td>9.643172</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1, 9)\n",
       "┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐\n",
       "│ repertoir ┆ total_seq ┆ unique_pr ┆ total_cou ┆ … ┆ gini_coef ┆ top_produ ┆ convergen ┆ shannon_ │\n",
       "│ e_id      ┆ uences    ┆ oductive_ ┆ nt        ┆   ┆ ficient   ┆ ctive_seq ┆ ce        ┆ diversit │\n",
       "│ ---       ┆ ---       ┆ sequences ┆ ---       ┆   ┆ ---       ┆ uence     ┆ ---       ┆ y        │\n",
       "│ str       ┆ i64       ┆ ---       ┆ i64       ┆   ┆ f64       ┆ ---       ┆ f64       ┆ ---      │\n",
       "│           ┆           ┆ i64       ┆           ┆   ┆           ┆ f64       ┆           ┆ f64      │\n",
       "╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡\n",
       "│ FIN_15625 ┆ 13688     ┆ 13541     ┆ 15353     ┆ … ┆ 0.101629  ┆ 0.001238  ┆ 1.010856  ┆ 9.643172 │\n",
       "│ _AgXP_TCR ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ B         ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_diversity_metrics(airr_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "271da121",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nucleotide</th>\n",
       "      <th>aminoAcid</th>\n",
       "      <th>count (templates/reads)</th>\n",
       "      <th>frequencyCount (%)</th>\n",
       "      <th>cdr3Length</th>\n",
       "      <th>vMaxResolved</th>\n",
       "      <th>vFamilyName</th>\n",
       "      <th>vGeneName</th>\n",
       "      <th>vGeneAllele</th>\n",
       "      <th>vFamilyTies</th>\n",
       "      <th>vGeneNameTies</th>\n",
       "      <th>vGeneAlleleTies</th>\n",
       "      <th>dMaxResolved</th>\n",
       "      <th>dFamilyName</th>\n",
       "      <th>dGeneName</th>\n",
       "      <th>dGeneAllele</th>\n",
       "      <th>dFamilyTies</th>\n",
       "      <th>dGeneNameTies</th>\n",
       "      <th>dGeneAlleleTies</th>\n",
       "      <th>jMaxResolved</th>\n",
       "      <th>jFamilyName</th>\n",
       "      <th>jGeneName</th>\n",
       "      <th>jGeneAllele</th>\n",
       "      <th>jFamilyTies</th>\n",
       "      <th>jGeneNameTies</th>\n",
       "      <th>jGeneAlleleTies</th>\n",
       "      <th>vDeletion</th>\n",
       "      <th>n1Insertion</th>\n",
       "      <th>d5Deletion</th>\n",
       "      <th>d3Deletion</th>\n",
       "      <th>n2Insertion</th>\n",
       "      <th>jDeletion</th>\n",
       "      <th>vIndex</th>\n",
       "      <th>n1Index</th>\n",
       "      <th>dIndex</th>\n",
       "      <th>n2Index</th>\n",
       "      <th>jIndex</th>\n",
       "      <th>estimatedNumberGenomes</th>\n",
       "      <th>sequenceStatus</th>\n",
       "      <th>cloneResolved</th>\n",
       "      <th>vOrphon</th>\n",
       "      <th>dOrphon</th>\n",
       "      <th>jOrphon</th>\n",
       "      <th>vFunction</th>\n",
       "      <th>dFunction</th>\n",
       "      <th>jFunction</th>\n",
       "      <th>fractionNucleated</th>\n",
       "      <th>vAlignLength</th>\n",
       "      <th>vAlignSubstitutionCount</th>\n",
       "      <th>vAlignSubstitutionIndexes</th>\n",
       "      <th>vAlignSubstitutionGeneThreePrimeIndexes</th>\n",
       "      <th>vSeqWithMutations</th>\n",
       "      <th>sample_short</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CACCTACACACCCTGCAGCCAGAAGACTCGGCCCTGTATCTCTGCG...</td>\n",
       "      <td>CASSQDKNQPQHF</td>\n",
       "      <td>14</td>\n",
       "      <td>0.000759</td>\n",
       "      <td>39</td>\n",
       "      <td>TCRBV04-03*01</td>\n",
       "      <td>TCRBV04</td>\n",
       "      <td>TCRBV04-03</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBD01,TCRBD02</td>\n",
       "      <td>TCRBD01-01,TCRBD02-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBJ01-05*01</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>TCRBJ01-05</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>42</td>\n",
       "      <td>59</td>\n",
       "      <td>61</td>\n",
       "      <td>-1</td>\n",
       "      <td>63</td>\n",
       "      <td>14</td>\n",
       "      <td>In</td>\n",
       "      <td>VDJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F15625E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATCCAGCGCACACAGCAGGAGGACTCGGCCGTGTATCTCTGTGCCA...</td>\n",
       "      <td>CASSLMSGPGELFF</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000271</td>\n",
       "      <td>42</td>\n",
       "      <td>TCRBV07-02</td>\n",
       "      <td>TCRBV07</td>\n",
       "      <td>TCRBV07-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>01,02</td>\n",
       "      <td>TCRBD02-01</td>\n",
       "      <td>TCRBD02</td>\n",
       "      <td>TCRBD02-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>01,02</td>\n",
       "      <td>TCRBJ02-02*01</td>\n",
       "      <td>TCRBJ02</td>\n",
       "      <td>TCRBJ02-02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>39</td>\n",
       "      <td>54</td>\n",
       "      <td>58</td>\n",
       "      <td>62</td>\n",
       "      <td>64</td>\n",
       "      <td>5</td>\n",
       "      <td>In</td>\n",
       "      <td>VDJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F15625E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCAGAACCCAGGGACTCAGCTGTGTACTTCTGTGCCAGCAGACGCC...</td>\n",
       "      <td>CASRRPTGQGIRSGYTF</td>\n",
       "      <td>9</td>\n",
       "      <td>0.000488</td>\n",
       "      <td>51</td>\n",
       "      <td>TCRBV12</td>\n",
       "      <td>TCRBV12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBV12-03/12-04,TCRBV12-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBD01-01*01</td>\n",
       "      <td>TCRBD01</td>\n",
       "      <td>TCRBD01-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBJ01-02*01</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>TCRBJ01-02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>30</td>\n",
       "      <td>41</td>\n",
       "      <td>50</td>\n",
       "      <td>56</td>\n",
       "      <td>69</td>\n",
       "      <td>9</td>\n",
       "      <td>In</td>\n",
       "      <td>VDJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F15625E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ATCAATTCCCTGGAGCTTGGTGACTCTGCTGTGTATTTCTGTGCCA...</td>\n",
       "      <td>CASSQGTGGSPLHF</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000434</td>\n",
       "      <td>42</td>\n",
       "      <td>TCRBV03-01/03-02*01</td>\n",
       "      <td>TCRBV03</td>\n",
       "      <td>TCRBV03-01/03-02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBD01-01*01</td>\n",
       "      <td>TCRBD01</td>\n",
       "      <td>TCRBD01-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBJ01-06*02</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>TCRBJ01-06</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>39</td>\n",
       "      <td>55</td>\n",
       "      <td>57</td>\n",
       "      <td>-1</td>\n",
       "      <td>65</td>\n",
       "      <td>8</td>\n",
       "      <td>In</td>\n",
       "      <td>VDJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F15625E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GATCCAGCGCACAGAGCAGGGGGACTCAGCTGTGTATCTCTGTGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000217</td>\n",
       "      <td>41</td>\n",
       "      <td>TCRBV07-04*01</td>\n",
       "      <td>TCRBV07</td>\n",
       "      <td>TCRBV07-04</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBD01-01*01</td>\n",
       "      <td>TCRBD01</td>\n",
       "      <td>TCRBD01-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBJ01-02*01</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>TCRBJ01-02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>40</td>\n",
       "      <td>53</td>\n",
       "      <td>56</td>\n",
       "      <td>61</td>\n",
       "      <td>64</td>\n",
       "      <td>4</td>\n",
       "      <td>Out</td>\n",
       "      <td>VDJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F15625E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16452</th>\n",
       "      <td>CATCAATTCCCTGGAGCTTGGTGACTCTGCTGTGTATTTCTGTGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>41</td>\n",
       "      <td>TCRBV03-01/03-02*01</td>\n",
       "      <td>TCRBV03</td>\n",
       "      <td>TCRBV03-01/03-02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBD02-01</td>\n",
       "      <td>TCRBD02</td>\n",
       "      <td>TCRBD02-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>01,02</td>\n",
       "      <td>TCRBJ02-02*01</td>\n",
       "      <td>TCRBJ02</td>\n",
       "      <td>TCRBJ02-02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>18</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>56</td>\n",
       "      <td>61</td>\n",
       "      <td>76</td>\n",
       "      <td>1</td>\n",
       "      <td>Out</td>\n",
       "      <td>VDJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F15625E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16453</th>\n",
       "      <td>AAGATCCAGCCTGCAGAGCTTGGGGACTCGGCCGTGTATCTCTGTG...</td>\n",
       "      <td>CASSLDPDSPAFF</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>39</td>\n",
       "      <td>TCRBV11-03*01</td>\n",
       "      <td>TCRBV11</td>\n",
       "      <td>TCRBV11-03</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBD02-01</td>\n",
       "      <td>TCRBD02</td>\n",
       "      <td>TCRBD02-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>01,02</td>\n",
       "      <td>TCRBJ01-01*01</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>TCRBJ01-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "      <td>42</td>\n",
       "      <td>59</td>\n",
       "      <td>63</td>\n",
       "      <td>67</td>\n",
       "      <td>71</td>\n",
       "      <td>1</td>\n",
       "      <td>In</td>\n",
       "      <td>VDJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F15625E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16454</th>\n",
       "      <td>CACATCAATTCCCTGGAGCTTGGTGACTCTGCTGTGTATTTCTGTG...</td>\n",
       "      <td>CASSLGGATEAFF</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>39</td>\n",
       "      <td>TCRBV03-01/03-02*01</td>\n",
       "      <td>TCRBV03</td>\n",
       "      <td>TCRBV03-01/03-02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBD02-01*01</td>\n",
       "      <td>TCRBD02</td>\n",
       "      <td>TCRBD02-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBJ01-01*01</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>TCRBJ01-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>42</td>\n",
       "      <td>55</td>\n",
       "      <td>56</td>\n",
       "      <td>64</td>\n",
       "      <td>65</td>\n",
       "      <td>1</td>\n",
       "      <td>In</td>\n",
       "      <td>VDJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F15625E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16455</th>\n",
       "      <td>AGTGACCAGTGCCCATCCTGAAGACAGCAGCTTCTACATCTGCAGT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>41</td>\n",
       "      <td>TCRBV20</td>\n",
       "      <td>TCRBV20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBV20-01,TCRBV20-or09_02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBD01,TCRBD02</td>\n",
       "      <td>TCRBD01-01,TCRBD02-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBJ01-01*01</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>TCRBJ01-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>40</td>\n",
       "      <td>47</td>\n",
       "      <td>52</td>\n",
       "      <td>57</td>\n",
       "      <td>65</td>\n",
       "      <td>1</td>\n",
       "      <td>Out</td>\n",
       "      <td>VDJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F15625E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16456</th>\n",
       "      <td>GGAGTCAGGGGACACAGCACTGTATTTCTGTGCCAGCAGCAACCTC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>53</td>\n",
       "      <td>TCRBV21-01*01</td>\n",
       "      <td>TCRBV21</td>\n",
       "      <td>TCRBV21-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBD01-01*01</td>\n",
       "      <td>TCRBD01</td>\n",
       "      <td>TCRBD01-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCRBJ01-01*01</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>TCRBJ01-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>42</td>\n",
       "      <td>48</td>\n",
       "      <td>58</td>\n",
       "      <td>62</td>\n",
       "      <td>1</td>\n",
       "      <td>Out</td>\n",
       "      <td>VDJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F15625E</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16457 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              nucleotide          aminoAcid  \\\n",
       "0      CACCTACACACCCTGCAGCCAGAAGACTCGGCCCTGTATCTCTGCG...      CASSQDKNQPQHF   \n",
       "1      ATCCAGCGCACACAGCAGGAGGACTCGGCCGTGTATCTCTGTGCCA...     CASSLMSGPGELFF   \n",
       "2      TCAGAACCCAGGGACTCAGCTGTGTACTTCTGTGCCAGCAGACGCC...  CASRRPTGQGIRSGYTF   \n",
       "3      ATCAATTCCCTGGAGCTTGGTGACTCTGCTGTGTATTTCTGTGCCA...     CASSQGTGGSPLHF   \n",
       "4      GATCCAGCGCACAGAGCAGGGGGACTCAGCTGTGTATCTCTGTGCC...                NaN   \n",
       "...                                                  ...                ...   \n",
       "16452  CATCAATTCCCTGGAGCTTGGTGACTCTGCTGTGTATTTCTGTGCC...                NaN   \n",
       "16453  AAGATCCAGCCTGCAGAGCTTGGGGACTCGGCCGTGTATCTCTGTG...      CASSLDPDSPAFF   \n",
       "16454  CACATCAATTCCCTGGAGCTTGGTGACTCTGCTGTGTATTTCTGTG...      CASSLGGATEAFF   \n",
       "16455  AGTGACCAGTGCCCATCCTGAAGACAGCAGCTTCTACATCTGCAGT...                NaN   \n",
       "16456  GGAGTCAGGGGACACAGCACTGTATTTCTGTGCCAGCAGCAACCTC...                NaN   \n",
       "\n",
       "       count (templates/reads)  frequencyCount (%)  cdr3Length  \\\n",
       "0                           14            0.000759          39   \n",
       "1                            5            0.000271          42   \n",
       "2                            9            0.000488          51   \n",
       "3                            8            0.000434          42   \n",
       "4                            4            0.000217          41   \n",
       "...                        ...                 ...         ...   \n",
       "16452                        1            0.000054          41   \n",
       "16453                        1            0.000054          39   \n",
       "16454                        1            0.000054          39   \n",
       "16455                        1            0.000054          41   \n",
       "16456                        1            0.000054          53   \n",
       "\n",
       "              vMaxResolved vFamilyName         vGeneName  vGeneAllele  \\\n",
       "0            TCRBV04-03*01     TCRBV04        TCRBV04-03          1.0   \n",
       "1               TCRBV07-02     TCRBV07        TCRBV07-02          NaN   \n",
       "2                  TCRBV12     TCRBV12               NaN          NaN   \n",
       "3      TCRBV03-01/03-02*01     TCRBV03  TCRBV03-01/03-02          1.0   \n",
       "4            TCRBV07-04*01     TCRBV07        TCRBV07-04          1.0   \n",
       "...                    ...         ...               ...          ...   \n",
       "16452  TCRBV03-01/03-02*01     TCRBV03  TCRBV03-01/03-02          1.0   \n",
       "16453        TCRBV11-03*01     TCRBV11        TCRBV11-03          1.0   \n",
       "16454  TCRBV03-01/03-02*01     TCRBV03  TCRBV03-01/03-02          1.0   \n",
       "16455              TCRBV20     TCRBV20               NaN          NaN   \n",
       "16456        TCRBV21-01*01     TCRBV21        TCRBV21-01          1.0   \n",
       "\n",
       "      vFamilyTies                vGeneNameTies vGeneAlleleTies   dMaxResolved  \\\n",
       "0             NaN                          NaN             NaN            NaN   \n",
       "1             NaN                          NaN           01,02     TCRBD02-01   \n",
       "2             NaN  TCRBV12-03/12-04,TCRBV12-04             NaN  TCRBD01-01*01   \n",
       "3             NaN                          NaN             NaN  TCRBD01-01*01   \n",
       "4             NaN                          NaN             NaN  TCRBD01-01*01   \n",
       "...           ...                          ...             ...            ...   \n",
       "16452         NaN                          NaN             NaN     TCRBD02-01   \n",
       "16453         NaN                          NaN             NaN     TCRBD02-01   \n",
       "16454         NaN                          NaN             NaN  TCRBD02-01*01   \n",
       "16455         NaN   TCRBV20-01,TCRBV20-or09_02             NaN            NaN   \n",
       "16456         NaN                          NaN             NaN  TCRBD01-01*01   \n",
       "\n",
       "      dFamilyName   dGeneName  dGeneAllele      dFamilyTies  \\\n",
       "0             NaN         NaN          NaN  TCRBD01,TCRBD02   \n",
       "1         TCRBD02  TCRBD02-01          NaN              NaN   \n",
       "2         TCRBD01  TCRBD01-01          1.0              NaN   \n",
       "3         TCRBD01  TCRBD01-01          1.0              NaN   \n",
       "4         TCRBD01  TCRBD01-01          1.0              NaN   \n",
       "...           ...         ...          ...              ...   \n",
       "16452     TCRBD02  TCRBD02-01          NaN              NaN   \n",
       "16453     TCRBD02  TCRBD02-01          NaN              NaN   \n",
       "16454     TCRBD02  TCRBD02-01          1.0              NaN   \n",
       "16455         NaN         NaN          NaN  TCRBD01,TCRBD02   \n",
       "16456     TCRBD01  TCRBD01-01          1.0              NaN   \n",
       "\n",
       "               dGeneNameTies dGeneAlleleTies   jMaxResolved jFamilyName  \\\n",
       "0      TCRBD01-01,TCRBD02-01             NaN  TCRBJ01-05*01     TCRBJ01   \n",
       "1                        NaN           01,02  TCRBJ02-02*01     TCRBJ02   \n",
       "2                        NaN             NaN  TCRBJ01-02*01     TCRBJ01   \n",
       "3                        NaN             NaN  TCRBJ01-06*02     TCRBJ01   \n",
       "4                        NaN             NaN  TCRBJ01-02*01     TCRBJ01   \n",
       "...                      ...             ...            ...         ...   \n",
       "16452                    NaN           01,02  TCRBJ02-02*01     TCRBJ02   \n",
       "16453                    NaN           01,02  TCRBJ01-01*01     TCRBJ01   \n",
       "16454                    NaN             NaN  TCRBJ01-01*01     TCRBJ01   \n",
       "16455  TCRBD01-01,TCRBD02-01             NaN  TCRBJ01-01*01     TCRBJ01   \n",
       "16456                    NaN             NaN  TCRBJ01-01*01     TCRBJ01   \n",
       "\n",
       "        jGeneName  jGeneAllele  jFamilyTies jGeneNameTies jGeneAlleleTies  \\\n",
       "0      TCRBJ01-05          1.0          NaN           NaN             NaN   \n",
       "1      TCRBJ02-02          1.0          NaN           NaN             NaN   \n",
       "2      TCRBJ01-02          1.0          NaN           NaN             NaN   \n",
       "3      TCRBJ01-06          2.0          NaN           NaN             NaN   \n",
       "4      TCRBJ01-02          1.0          NaN           NaN             NaN   \n",
       "...           ...          ...          ...           ...             ...   \n",
       "16452  TCRBJ02-02          1.0          NaN           NaN             NaN   \n",
       "16453  TCRBJ01-01          1.0          NaN           NaN             NaN   \n",
       "16454  TCRBJ01-01          1.0          NaN           NaN             NaN   \n",
       "16455  TCRBJ01-01          1.0          NaN           NaN             NaN   \n",
       "16456  TCRBJ01-01          1.0          NaN           NaN             NaN   \n",
       "\n",
       "       vDeletion  n1Insertion  d5Deletion  d3Deletion  n2Insertion  jDeletion  \\\n",
       "0              0            2           5           5            0          4   \n",
       "1              2            4           8           4            2          6   \n",
       "2              6            9           0           6           13          8   \n",
       "3              1            2           3           1            0          9   \n",
       "4              4            3           2           5            3          3   \n",
       "...          ...          ...         ...         ...          ...        ...   \n",
       "16452          7            6           8           3           15         18   \n",
       "16453          0            4           2          10            4         10   \n",
       "16454          4            1           8           0            1          4   \n",
       "16455          7            5           5           2            8          4   \n",
       "16456          3            6           2           0            4          1   \n",
       "\n",
       "       vIndex  n1Index  dIndex  n2Index  jIndex  estimatedNumberGenomes  \\\n",
       "0          42       59      61       -1      63                      14   \n",
       "1          39       54      58       62      64                       5   \n",
       "2          30       41      50       56      69                       9   \n",
       "3          39       55      57       -1      65                       8   \n",
       "4          40       53      56       61      64                       4   \n",
       "...       ...      ...     ...      ...     ...                     ...   \n",
       "16452      40       50      56       61      76                       1   \n",
       "16453      42       59      63       67      71                       1   \n",
       "16454      42       55      56       64      65                       1   \n",
       "16455      40       47      52       57      65                       1   \n",
       "16456      28       42      48       58      62                       1   \n",
       "\n",
       "      sequenceStatus cloneResolved  vOrphon  dOrphon  jOrphon  vFunction  \\\n",
       "0                 In           VDJ      NaN      NaN      NaN        NaN   \n",
       "1                 In           VDJ      NaN      NaN      NaN        NaN   \n",
       "2                 In           VDJ      NaN      NaN      NaN        NaN   \n",
       "3                 In           VDJ      NaN      NaN      NaN        NaN   \n",
       "4                Out           VDJ      NaN      NaN      NaN        NaN   \n",
       "...              ...           ...      ...      ...      ...        ...   \n",
       "16452            Out           VDJ      NaN      NaN      NaN        NaN   \n",
       "16453             In           VDJ      NaN      NaN      NaN        NaN   \n",
       "16454             In           VDJ      NaN      NaN      NaN        NaN   \n",
       "16455            Out           VDJ      NaN      NaN      NaN        NaN   \n",
       "16456            Out           VDJ      NaN      NaN      NaN        NaN   \n",
       "\n",
       "       dFunction  jFunction  fractionNucleated  vAlignLength  \\\n",
       "0            NaN        NaN                NaN           NaN   \n",
       "1            NaN        NaN                NaN           NaN   \n",
       "2            NaN        NaN                NaN           NaN   \n",
       "3            NaN        NaN                NaN           NaN   \n",
       "4            NaN        NaN                NaN           NaN   \n",
       "...          ...        ...                ...           ...   \n",
       "16452        NaN        NaN                NaN           NaN   \n",
       "16453        NaN        NaN                NaN           NaN   \n",
       "16454        NaN        NaN                NaN           NaN   \n",
       "16455        NaN        NaN                NaN           NaN   \n",
       "16456        NaN        NaN                NaN           NaN   \n",
       "\n",
       "       vAlignSubstitutionCount  vAlignSubstitutionIndexes  \\\n",
       "0                          NaN                        NaN   \n",
       "1                          NaN                        NaN   \n",
       "2                          NaN                        NaN   \n",
       "3                          NaN                        NaN   \n",
       "4                          NaN                        NaN   \n",
       "...                        ...                        ...   \n",
       "16452                      NaN                        NaN   \n",
       "16453                      NaN                        NaN   \n",
       "16454                      NaN                        NaN   \n",
       "16455                      NaN                        NaN   \n",
       "16456                      NaN                        NaN   \n",
       "\n",
       "       vAlignSubstitutionGeneThreePrimeIndexes  vSeqWithMutations sample_short  \n",
       "0                                          NaN                NaN      F15625E  \n",
       "1                                          NaN                NaN      F15625E  \n",
       "2                                          NaN                NaN      F15625E  \n",
       "3                                          NaN                NaN      F15625E  \n",
       "4                                          NaN                NaN      F15625E  \n",
       "...                                        ...                ...          ...  \n",
       "16452                                      NaN                NaN      F15625E  \n",
       "16453                                      NaN                NaN      F15625E  \n",
       "16454                                      NaN                NaN      F15625E  \n",
       "16455                                      NaN                NaN      F15625E  \n",
       "16456                                      NaN                NaN      F15625E  \n",
       "\n",
       "[16457 rows x 53 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70ce8fa8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vFamilyName</th>\n",
       "      <th>jGeneName</th>\n",
       "      <th>cdr3Length</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>TCRBV05</td>\n",
       "      <td>TCRBJ01-01</td>\n",
       "      <td>42</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658</th>\n",
       "      <td>TCRBV05</td>\n",
       "      <td>TCRBJ02-01</td>\n",
       "      <td>45</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>827</th>\n",
       "      <td>TCRBV06</td>\n",
       "      <td>TCRBJ01-02</td>\n",
       "      <td>42</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>800</th>\n",
       "      <td>TCRBV06</td>\n",
       "      <td>TCRBJ01-01</td>\n",
       "      <td>42</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>TCRBV05</td>\n",
       "      <td>TCRBJ01-02</td>\n",
       "      <td>42</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    vFamilyName   jGeneName  cdr3Length  count\n",
       "535     TCRBV05  TCRBJ01-01          42     80\n",
       "658     TCRBV05  TCRBJ02-01          45     72\n",
       "827     TCRBV06  TCRBJ01-02          42     70\n",
       "800     TCRBV06  TCRBJ01-01          42     67\n",
       "564     TCRBV05  TCRBJ01-02          42     66"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a table summarizing the usage of vFamilyName, jGeneName, and cdr3Length\n",
    "usage_summary = sample_df.groupby(['vFamilyName', 'jGeneName', 'cdr3Length']).size().reset_index(name='count')\n",
    "\n",
    "# Arrange the table from highest to lowest count\n",
    "usage_summary = usage_summary.sort_values(by='count', ascending=False)\n",
    "\n",
    "# Display the table\n",
    "usage_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90b05232",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vFamilyName</th>\n",
       "      <th>jFamilyName</th>\n",
       "      <th>cdr3Length</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>TCRBV05</td>\n",
       "      <td>TCRBJ02</td>\n",
       "      <td>45</td>\n",
       "      <td>271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>TCRBV05</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>42</td>\n",
       "      <td>249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>TCRBV06</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>42</td>\n",
       "      <td>243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>TCRBV05</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>45</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>TCRBV05</td>\n",
       "      <td>TCRBJ02</td>\n",
       "      <td>42</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    vFamilyName jFamilyName  cdr3Length  count\n",
       "244     TCRBV05     TCRBJ02          45    271\n",
       "202     TCRBV05     TCRBJ01          42    249\n",
       "277     TCRBV06     TCRBJ01          42    243\n",
       "205     TCRBV05     TCRBJ01          45    222\n",
       "241     TCRBV05     TCRBJ02          42    219"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a table summarizing the usage of vFamilyName, jGeneName, and cdr3Length\n",
    "usage_summary = sample_df.groupby(['vFamilyName', 'jFamilyName', 'cdr3Length']).size().reset_index(name='count')\n",
    "\n",
    "# Arrange the table from highest to lowest count\n",
    "usage_summary = usage_summary.sort_values(by='count', ascending=False)\n",
    "\n",
    "# Display the table\n",
    "usage_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52391a58",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4162f4ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        group  row_count  row_fraction  cell_count  cell_fraction  \\\n",
      "0     TCRBV05       2063      0.125502        2326       0.126180   \n",
      "1     TCRBV06       2031      0.123555        2265       0.122871   \n",
      "2     TCRBV07       1950      0.118628        2226       0.120755   \n",
      "3     TCRBV20       1314      0.079937        1447       0.078496   \n",
      "4     TCRBV28       1019      0.061991        1144       0.062059   \n",
      "...       ...        ...           ...         ...            ...   \n",
      "1422      NaN          5      0.000304           6       0.000325   \n",
      "1423      NaN          6      0.000365           6       0.000325   \n",
      "1424      NaN          1      0.000061           1       0.000054   \n",
      "1425      NaN          1      0.000061           1       0.000054   \n",
      "1426      NaN          1      0.000061           1       0.000054   \n",
      "\n",
      "                                     param vFamilyName jFamilyName  cdr3Length  \n",
      "0                              vFamilyName         NaN         NaN         NaN  \n",
      "1                              vFamilyName         NaN         NaN         NaN  \n",
      "2                              vFamilyName         NaN         NaN         NaN  \n",
      "3                              vFamilyName         NaN         NaN         NaN  \n",
      "4                              vFamilyName         NaN         NaN         NaN  \n",
      "...                                    ...         ...         ...         ...  \n",
      "1422  vFamilyName, jFamilyName, cdr3Length     TCRBV30     TCRBJ02        48.0  \n",
      "1423  vFamilyName, jFamilyName, cdr3Length     TCRBV30     TCRBJ02        51.0  \n",
      "1424  vFamilyName, jFamilyName, cdr3Length     TCRBV30     TCRBJ02        54.0  \n",
      "1425  vFamilyName, jFamilyName, cdr3Length     TCRBV30     TCRBJ02        55.0  \n",
      "1426  vFamilyName, jFamilyName, cdr3Length     TCRBV30     TCRBJ02        59.0  \n",
      "\n",
      "[1427 rows x 9 columns]\n"
     ]
    }
   ],
   "source": [
    "# Compute summaries for vFamilyName, jFamilyName, cdr3Length, and jGeneName\n",
    "summary_params = ['vFamilyName', 'jFamilyName', 'cdr3Length', 'jGeneName']\n",
    "all_summaries = []\n",
    "\n",
    "for param in summary_params:\n",
    "    # Row count and fraction\n",
    "    param_summary = sample_df[param].value_counts().reset_index()\n",
    "    param_summary.columns = ['group', 'row_count']\n",
    "    param_summary['row_fraction'] = param_summary['row_count'] / param_summary['row_count'].sum()\n",
    "\n",
    "    # Cell count and fraction based on 'count (templates/reads)'\n",
    "    cell_counts = sample_df.groupby(param)['count (templates/reads)'].sum().reset_index()\n",
    "    cell_counts.columns = [param, 'cell_count']\n",
    "    cell_counts['cell_fraction'] = cell_counts['cell_count'] / cell_counts['cell_count'].sum()\n",
    "\n",
    "    # Merge summaries\n",
    "    param_summary = param_summary.merge(cell_counts, left_on='group', right_on=param).drop(columns=[param])\n",
    "\n",
    "    # Add parameter column\n",
    "    param_summary['param'] = param\n",
    "\n",
    "    # Append to list\n",
    "    all_summaries.append(param_summary)\n",
    "\n",
    "# Concatenate all summaries\n",
    "final_summary = pd.concat(all_summaries, ignore_index=True)\n",
    "\n",
    "# Compute summary for grouping by ['vFamilyName', 'jFamilyName', 'cdr3Length']\n",
    "grouped_summary = sample_df.groupby(['vFamilyName', 'jFamilyName', 'cdr3Length']).agg(\n",
    "    row_count=('vFamilyName', 'size'),\n",
    "    cell_count=('count (templates/reads)', 'sum')\n",
    ").reset_index()\n",
    "\n",
    "# Add row_fraction and cell_fraction\n",
    "grouped_summary['row_fraction'] = grouped_summary['row_count'] / grouped_summary['row_count'].sum()\n",
    "grouped_summary['cell_fraction'] = grouped_summary['cell_count'] / grouped_summary['cell_count'].sum()\n",
    "\n",
    "# Add parameter column\n",
    "grouped_summary['param'] = 'vFamilyName, jFamilyName, cdr3Length'\n",
    "\n",
    "# Append grouped summary to final summary\n",
    "final_summary = pd.concat([final_summary, grouped_summary], ignore_index=True)\n",
    "\n",
    "# Display the final summary\n",
    "return final_summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecd1979d",
   "metadata": {},
   "source": [
    "sample_df.type "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "8ad9ed52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.DataFrame"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(sample_df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
